crates/unftp-auth-json/src/lib.rs

```diff
@@ -0,0 +1,168 @@
+//! [`Authenticator`] implementation that authenticates against a JSON file.
+//!
+//! [`Authenticator`]: libunftp::auth::Authenticator
+
+use async_trait::async_trait;
+use libunftp::auth::{AuthenticationError, Authenticator, DefaultUser};
+use ring::{
+    digest::SHA512_OUTPUT_LEN,
+    pbkdf2::{verify, PBKDF2_HMAC_SHA512},
+};
+use serde::Deserialize;
+use std::{
+    collections::{BTreeSet, HashMap},
+    convert::TryInto,
+    fs,
+    num::NonZeroU32,
+    path::Path,
+    time::Duration,
+};
+use tokio::time::sleep;
+
+#[derive(Deserialize, Clone, Debug)]
+struct Credential {
+    username: String,
+    pbkdf2_salt: String,
+    pbkdf2_key: String,
+    pbkdf2_iter: NonZeroU32,
+}
+
+/// [`Authenticator`](libunftp::auth::Authenticator) implementation that authenticates against a JSON.
+///
+/// Verifies passwords against pbkdf2_key using the corresponding parameters form JSON.
+/// Example credentials file format:
+/// [
+//   {
+//     "username": "testuser1",
+//     "pbkdf2_salt": "testuser1.acme.bol.com",
+//     "pbkdf2_key": "<<BASE_64_KDF>>",
+//     "pbkdf2_iter": 500000
+//   },
+//   {
+//     "username": "testuser2",
+//     "pbkdf2_salt": "testuser2.acme.bol.com",
+//     "pbkdf2_key": "<<BASE_64_KDF>>",
+//     "pbkdf2_iter": 500000
+//   }
+// ]
+
+#[derive(Clone, Debug)]
+pub struct JsonAuthenticator {
+    db: HashMap<String, Password>,
+}
+
+#[derive(Clone, Debug)]
+struct Password {
+    pbkdf2_salt: String,
+    pbkdf2_key: [u8; SHA512_OUTPUT_LEN],
+    pbkdf2_iter: NonZeroU32,
+}
+
+impl JsonAuthenticator {
+    /// Initialize a new [`JsonFileAuthenticator`] from file.
+    pub fn from_file<P: AsRef<Path>>(filename: P) -> Result<Self, Box<dyn std::error::Error>> {
+        let json: String = fs::read_to_string(filename)?;
+
+        JsonAuthenticator::from_json(json)
+    }
+
+    /// Initialize a new [`JsonFileAuthenticator`] from json string.
+    pub fn from_json<T: Into<String>>(json: T) -> Result<Self, Box<dyn std::error::Error>> {
+        let db: Vec<Credential> = serde_json::from_str::<Vec<Credential>>(&json.into())?;
+        let salts: BTreeSet<String> = db.iter().map(|credential| credential.pbkdf2_salt.clone()).collect();
+        if db.len() != salts.len() {
+            return Err(Box::new(std::io::Error::from(std::io::ErrorKind::InvalidData)));
```

Comment:
Error could be made more user friendly, to make clear that the cause is the multiple use of a particular salt value. I'd prefer a warning over a failure on this however.

Reply:
Changed to `AuthenticationError::new("The provided salts for the JsonFileAuthenticator must be unique.")`